Expose schema endpoints

azrikahar opened the pull request:
## Description

Closes #11481, fixes #14687
Fixes ENG-221, ENG-270
- [x] Expose new `/schema/snapshot`, `/schema/apply`, `/schema/diff` API endpoints
- [x] Add export as YAML
- [x] Update specs
- [x] https://github.com/directus/directus/pull/14713/commits/e906f68710c647b97cfc85d7a93c7c5fd048fdae should be removed when https://github.com/knex/knex-schema-inspector/pull/129 is released in an updated package.
- [x] Verify #14687 is fixed
- [x] Update `knex-schema-inspector` to fix CrDB auto-increment detection https://github.com/knex/knex-schema-inspector/pull/135 
- [x] Revert changes to `.github/workflows/blackbox-main.yml`

## Type of Change

- [ ] Bugfix
- [ ] Improvement
- [x] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [x] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [x] Documentation was added/updated -> directus/docs#62

azrikahar committed:
Expose schema endpoints

azrikahar committed:
Merge branch 'main' into issue/11481

azrikahar committed:
respond with 204

azrikahar committed:
fix payload for export flag

azrikahar committed:
Merge branch 'issue/11481' of https://github.com/directus/directus into issue/11481

azrikahar committed:
allow export as YAML

azrikahar committed:
use SchemaService

azrikahar committed:
move getStringFromStream to utils

azrikahar committed:
clearer exceptions with logs

azrikahar committed:
check empty request body

azrikahar committed:
Merge branch 'main' into issue/11481

azrikahar committed:
specs

rijkvanzanten committed:
Merge branch 'main' into issue/11481

azrikahar committed:
Merge branch 'main' into issue/11481

rijkvanzanten committed:
Merge branch 'main' into issue/11481

rijkvanzanten commented on api/src/services/schema.ts:
```diff
@@ -0,0 +1,73 @@
+	async snapshot(): Promise<Snapshot> {
+		if (this.accountability?.admin !== true) throw new ForbiddenException();
+
+		await flushCaches();
```
Yeah good one. There's a bigger change we have to do in caches and Redis use in general, so I think we should do a bigger change across all of that in a separate sprint 👍🏻

azrikahar committed:
Merge branch 'main' into issue/11481

rijkvanzanten commented:
@licitdev Could you help with the blackbox tests for these new endpoints? 🙂

azrikahar committed:
Merge branch 'main' into issue/11481

licitdev committed:
Merge branch 'main' into issue/11481

licitdev commented on api/src/controllers/server.ts:
```diff
@@ -81,4 +88,113 @@ router.get(
+router.get(
+	'/schema/snapshot',
+	asyncHandler(async (req, res, next) => {
+		const service = new SchemaService({ accountability: req.accountability, schema: req.schema });
```
We should not pass `schema` in so as to fetch a fresh schema that bypasses cache.

```suggestion
		const service = new SchemaService({ accountability: req.accountability });
```

licitdev commented on api/src/controllers/server.ts:
```diff
@@ -81,4 +88,113 @@ router.get(
+	'/schema/apply',
+	asyncHandler(schemaMultipartHandler),
+	asyncHandler(async (req, res, next) => {
+		const service = new SchemaService({ accountability: req.accountability, schema: req.schema });
```
```suggestion
		const service = new SchemaService({ accountability: req.accountability });
```

licitdev commented on api/src/controllers/server.ts:
```diff
@@ -81,4 +88,113 @@ router.get(
+	'/schema/diff',
+	asyncHandler(schemaMultipartHandler),
+	asyncHandler(async (req, res, next) => {
+		const service = new SchemaService({ accountability: req.accountability, schema: req.schema });
```
```suggestion
		const service = new SchemaService({ accountability: req.accountability });
```

licitdev commented on api/src/services/schema.ts:
```diff
@@ -0,0 +1,73 @@
+	}
+
+	async snapshot(): Promise<Snapshot> {
+		if (this.accountability?.admin !== true) throw new ForbiddenException();
+
+		await flushCaches();
+
+		const currentSnapshot = await getSnapshot({ database: this.knex, schema: this.schema });
```
There is no need to `flushCache()` as we're fetching a fresh schema that bypasses cache.

```suggestion
		if (this.accountability?.admin !== true) throw new ForbiddenException();

		const currentSnapshot = await getSnapshot({ database: this.knex, schema: this.schema });
```

licitdev commented on api/src/services/schema.ts:
```diff
@@ -0,0 +1,73 @@
+	}
+
+	async apply(snapshot: Snapshot): Promise<void> {
+		if (this.accountability?.admin !== true) throw new ForbiddenException();
+
+		await flushCaches();
+
+		const currentSnapshot = await getSnapshot({ database: this.knex, schema: this.schema });
```
```suggestion
		if (this.accountability?.admin !== true) throw new ForbiddenException();

		const currentSnapshot = await getSnapshot({ database: this.knex, schema: this.schema });
```

licitdev commented on api/src/services/schema.ts:
```diff
@@ -0,0 +1,73 @@
+	}
+
+	async diff(snapshot: Snapshot): Promise<SnapshotDiff | null> {
+		if (this.accountability?.admin !== true) throw new ForbiddenException();
+
+		await flushCaches();
+
+		const currentSnapshot = await getSnapshot({ database: this.knex, schema: this.schema });
```
```suggestion
		if (this.accountability?.admin !== true) throw new ForbiddenException();

		const currentSnapshot = await getSnapshot({ database: this.knex, schema: this.schema });
```

licitdev reviewed (changes requested):
The schema that's currently passed in has been fetched in the `req.schema` middleware and it might be outdated as the cache is flushed after. In addition, I think we should reduce the flushing of cache.

Please refer to the following changes in addition to those that I have suggested.

The `getSchema()` can be updated to accept a `bypassCache` flag.
 https://github.com/directus/directus/blob/5895ced76c1d98055ed3112c2ef521f407342808/api/src/utils/get-schema.ts#L17-L20

The following snapshot utils should then be updated to call `getSchema({ database, bypassCache: true })`.

https://github.com/directus/directus/blob/5895ced76c1d98055ed3112c2ef521f407342808/api/src/utils/get-snapshot.ts#L12

https://github.com/directus/directus/blob/5895ced76c1d98055ed3112c2ef521f407342808/api/src/utils/apply-snapshot.ts#L24

Lastly, the flushing of cache from CLI should be removed as well.

https://github.com/directus/directus/blob/5895ced76c1d98055ed3112c2ef521f407342808/api/src/cli/commands/schema/snapshot.ts#L14

https://github.com/directus/directus/blob/5895ced76c1d98055ed3112c2ef521f407342808/api/src/cli/commands/schema/apply.ts#L22

azrikahar committed:
add bypassCache flag and remove flushCaches calls

azrikahar committed:
Merge branch 'main' into issue/11481

licitdev committed:
Identify integer typed fields in Oracle

licitdev committed:
Fix uuid foreign keys type for Oracle

licitdev committed:
Detect autoincrement for CockroachDB

licitdev committed:
Bypass cache for fields and relations services

licitdev committed:
Merge branch 'main' into issue/11481

licitdev committed:
Add same DB tests

licitdev committed:
Test if working for all vendors

licitdev committed:
Merge branch 'main' into issue/11481

azrikahar committed:
Merge branch 'main' into issue/11481

licitdev committed:
Add schema snapshot testing into sequential flow

azrikahar committed:
Merge branch 'main' into issue/11481

azrikahar committed:
Fix schema re-apply attempts when it's a different database vendor (#14816)

* prevent diffing on certain properties

* fix apply-snapshot test

* prevent updates to autoincrement fields' schema

* sanitize relations

azrikahar committed:
Merge branch 'main' into issue/11481

azrikahar committed:
remove `colleciton.schema.sql` parsing in test

It is now being sanitized by `sanitizeCollection()` function in get-snapshot-diff.ts

licitdev committed:
Change concurrency group to not cancel test running on main

azrikahar committed:
Merge branch 'main' into issue/11481

azrikahar committed:
remove multipart for apply & add hash checking

azrikahar committed:
check incoming version & add force query param

azrikahar committed:
refine error message & stub for nested hash

azrikahar committed:
add `vendor` top-level property

as an additional safety net for cross db vendor applies

azrikahar committed:
sanitize generated schema snapshots

azrikahar committed:
snapshot joi validation

azrikahar committed:
Merge branch 'main' into issue/11481

azrikahar committed:
diff joi validation

azrikahar committed:
minor cleanup

azrikahar committed:
extract applyDiff & use deep-diff applyChange

azrikahar committed:
use applyDiff in schema service

licitdev committed:
Mark vendor as optional

azrikahar committed:
Merge branch 'main' into issue/11481

licitdev committed:
Update tests to apply diffs

azrikahar committed:
move force flag check into service

licitdev committed:
Patch mssql vendor typo

licitdev committed:
Set relation no action in cyclic relations

licitdev committed:
Update mysql timestamp default value

licitdev committed:
Oracle cannot define no action

licitdev committed:
Update oracle timestamp type default values

azrikahar committed:
add hash checking for outdated diffs

azrikahar committed:
fix diff service & endpoint

licitdev committed:
Add hashes when returning after

licitdev committed:
Fix self referencing relations for Oracle

licitdev committed:
Add temp fix for CURRENT_TIMESTAMP defaults in Oracle

azrikahar committed:
clean up driver and database client types

licitdev commented:
https://github.com/directus/directus/pull/14713/commits/e906f68710c647b97cfc85d7a93c7c5fd048fdae should be removed when https://github.com/knex/knex-schema-inspector/pull/129 is merged.

azrikahar committed:
only require diff hash if not kind N

azrikahar committed:
update hash comparison logic for create/delete

licitdev committed:
Set no action constraint for self referencing M2M in MSSQL

licitdev committed:
Add basic hash check tests

azrikahar committed:
omit default value when hashing auto increments

Specifically for CockroachDB with differing sequences

azrikahar committed:
add vendor check

azrikahar committed:
update specs

rijkvanzanten committed:
Merge branch 'main' into issue/11481

ConnorSimply commented on api/src/services/schema.ts:
```diff
@@ -0,0 +1,288 @@
+const snapshotJoiSchema = Joi.object({
+	version: Joi.number().required(),
+	directus: Joi.string().required(),
+	vendor: Joi.string().valid('mysql', 'postgres', 'cockroachdb', 'sqlite', 'oracle', 'mssql', 'redshift').optional(),
```
Should this Joi validation pull the list of vendors from the Database type definition instead? Pardon me if I am way off the mark, I am still getting acquainted with everything!

licitdev replied:
Yes we should! I've just added type validation to the vendor list. Thanks! 👍

licitdev committed:
Merge branch 'main' into issue/11481

licitdev committed:
Validate vendors with type definition

licitdev committed:
Spread the vendors input array

ConnorSimply commented:
This all looks good to me, however, just a UX/DX thing I am curious about. Currently it returns the schema or diff inside the the returned JSON data property `{ "data": {SCHEMA}}`, it would be nice if it just returned the schema/diff at the root level so that you could in theory download the file and upload it without having to get the children data. @rijkvanzanten @licitdev

connorwinston reviewed (approved):
This all looks good to me, however, just a UX/DX thing I am curious about. Currently it returns the schema or diff inside the the returned JSON data property `{ "data": {SCHEMA}}`, it would be nice if it just returned the schema/diff at the root level so that you could in theory download the file and upload it without having to get the children data. @rijkvanzanten @licitdev

licitdev committed:
Merge branch 'main' into issue/11481

licitdev commented:
@ConnorSimply Nice train of thought! 👍 The `export` query parameter is supported and it allows for a direct download in the selected file format. This info had been added in https://github.com/directus/docs/pull/62. Here's a docs preview link for easy reference https://deploy-preview-62--directus-docs.netlify.app/reference/system/server.html#retrieve-schema-snapshot.

Additionally, a UI within the data studio will be added in the future which should improve the UX of these schema endpoints.

ConnorSimply commented:
> @ConnorSimply Nice train of thought! 👍 The `export` query parameter is supported and it allows for a direct download in the selected file format. This info had been added in [directus/docs#62](https://github.com/directus/docs/pull/62). Here's a docs preview link for easy reference https://deploy-preview-62--directus-docs.netlify.app/reference/system/server.html#retrieve-schema-snapshot.
> 
> Additionally, a UI within the data studio will be added in the future which should improve the UX of these schema endpoints.

Yeah, I saw the export query param on `snapshot` in the docs but not on `diff` so I wasn't for sure if it was available for that endpoint. I think the docs should make a clear point that it is recommended to use the export param on both endpoints because I can already imagine the discussions coming in about why you cant just upload the snapshot endpoint result to diff/apply.

rijkvanzanten commented:
I get what you're saying @ConnorSimply, but in this case I think consistency wins over the added DX for this one endpoint. Every endpoint in the API returns the data payload under `data`, so we should do that here as well. That being said, I'm down to open the floor for discussion to drop that `data` top level scope completely API-wide for a next major release 🙂

azrikahar commented:
> Currently it returns the schema or diff inside the the returned JSON data property `{ "data": {SCHEMA}}`, it would be nice if it just returned the schema/diff at the root level

It was originally returning it at the root level when it was first being developed! However currently the export query does assume it to always be in `res.locals.payload.data` rather than root level `res.locals.payload`, so export won't work if we don't nest it inside `.data`:

https://github.com/directus/directus/blob/dbfedb56fe2efe44187abfd31165faaa79ae7780/api/src/middleware/respond.ts#L63-L79

and since it would be a much bigger (and breaking) API wide change as Rijk mentioned, it was opted not  to be done in this PR 👍

rijkvanzanten committed:
Merge branch 'main' into issue/11481

azrikahar committed:
Merge branch 'main' into issue/11481

azrikahar committed:
re-add Export component

azrikahar committed:
re-add js-yaml to root

licitdev committed:
Propagate mutation options for schema apply

licitdev committed:
Verify that snapshots differ and clear cache before hash tests

licitdev committed:
Fix unit test

licitdev committed:
Revert temp fix for CURRENT_TIMESTAMP defaults in Oracle

azrikahar committed:
Merge branch 'main' into issue/11481

rijkvanzanten committed:
Merge branch 'main' into issue/11481

azrikahar committed:
Merge branch 'main' into issue/11481

paescuj commented on api/src/cli/utils/drivers.ts:
```diff
@@ -1,4 +1,6 @@
-export const drivers = {
+import { Driver } from '../../types';
+
+export const drivers: Record<Driver, string> = {
```
Sorry, a bit unrelated but WDYT about usuing a `Map` here instead?

```ts
const driversMap: Map<string, Driver> = new Map([
	['PostgreSQL / Redshift', 'pg'],
	['CockroachDB (Beta)', 'cockroachdb'],
	['MySQL / MariaDB / Aurora', 'mysql'],
	['SQLite', 'sqlite3'],
	['Microsoft SQL Server', 'mssql'],
	['Oracle Database', 'oracledb'],
]);

// For "api/src/cli/commands/init/index.ts"
export const drivers = Array.from(drivers.keys());
```

This way the `getDriverForClient` function could be as simple as:
```ts
return driversMap.get(client) || null
```

---

This import here

https://github.com/directus/directus/blob/5ac28ab28b085063f2dc96eccfa5d979867691f2/api/src/cli/utils/create-env/index.ts#L7 should probably changed to the following anyway:
```ts
import { Driver } from '../../../types';
```

rijkvanzanten replied:
Theoretically yes, but those `drivers` are also imported and used in other files, so I don't want to force that refactor through as part of this PR 👍🏻

paescuj committed:
Define and reuse type for export format

paescuj committed:
Define and reuse list of database clients

rijkvanzanten committed:
Merge branch 'main' into issue/11481

rijkvanzanten committed:
Merge branch 'main' into issue/11481

rijkvanzanten commented on api/src/services/schema.ts:
```diff
@@ -0,0 +1,290 @@
+
+	constructor(options: Omit<AbstractServiceOptions, 'schema'>) {
+		this.knex = options.knex || getDatabase();
+		this.accountability = options.accountability || null;
```
Lets use `??` for defaults like this to avoid accidental falsey accidents

azrikahar replied:
Updated!

These lines are technically referred from other existing services, so I assume we should update them as well? Perhaps another PR?

paescuj replied:
> These lines are technically referred from other existing services, so I assume we should update them as well? Perhaps another PR?

That's right! Yes, let's do it in another PR this one is already big 😄🎉

rijkvanzanten commented on api/src/services/schema.ts:
```diff
@@ -0,0 +1,290 @@
+					const matchingCollection = snapshotWithHash.collections.find((c) => c.hash === diffCollection.hash);
+					if (!matchingCollection) {
+						throw new InvalidPayloadException(
+							`Provided diff to update collection "${collection}" does not match the current instance's collection, indicating it has changed after this diff was generated. Please regenerate a new diff and try again.`
```
> Please regenerate a new diff [...]

This should be _Please generate a new diff_ or _Please regenerate the diff_

azrikahar replied:
Changed it to `Please generate a new diff` 👍

rijkvanzanten commented on api/src/services/schema.ts:
```diff
@@ -0,0 +1,290 @@
+				if (diffField.diff[0]?.kind === 'N') {
+					const existingField = snapshotWithHash.fields.find(
+						(f) => f.collection === diffField.collection && f.field === diffField.field
+					);
+					if (existingField) {
```
Tiny tiny nitpick, but I personally prefer putting an empty line between blocks like these. Makes the whole thing just a tad more readable

rijkvanzanten commented on api/src/services/schema.ts:
```diff
@@ -0,0 +1,290 @@
+							`Provided diff is trying to create field "${field}" but it already exists. Please regenerate a new diff and try again.`
+						);
+					}
+				} else if (diffField.diff[0]?.kind === 'D') {
```
I'm wondering if we should put the string values here in an enum so it's a little easier to read, like:

```
if (kind === Kind.DELETE)
```

azrikahar replied:
That's a great idea 👍

rijkvanzanten commented on api/src/services/schema.ts:
```diff
@@ -0,0 +1,290 @@
+
+		const snapshotWithHash = this.getHashedSnapshot(currentSnapshot);
+
+		if (payload.hash !== snapshotWithHash.hash) {
+			for (const diffCollection of payload.diff.collections) {
```
Maybe we should move this bit to a new method `validateDiff` so it's easier to test in separation

paescuj replied:
Done in [7217049](https://github.com/directus/directus/pull/14713/commits/72170492785f0d061aa11173bf1ccd45cd02fcb3). @rijkvanzanten Is this what you had in mind?

azrikahar replied:
@paescuj Thanks for making the changes! Thoughts on making `validateDiff` a void function like `validateSnapshot`? Just curious whether we should make them behave a tad more consistent (since both are validating stuff in schema service's methods)

paescuj replied:
Thanks for your feedback, @azrikahar! Absolutely, we should 😃 I'm going to update it according to your suggestions 👍

rijkvanzanten replied:
LGTM as is 👍🏻

rijkvanzanten commented on api/src/services/schema.ts:
```diff
@@ -0,0 +1,290 @@
+			);
+		}
+
+		if (
+			payload.diff.collections.length === 0 &&
+			payload.diff.fields.length === 0 &&
+			payload.diff.relations.length === 0
+		) {
+			return;
+		}
```
Should this check be first, so we can skip the validation step if there's no diff to be applied?

rijkvanzanten commented on api/src/services/schema.ts:
```diff
@@ -0,0 +1,290 @@
+		const snapshotHash = getVersionedHash(snapshot);
+
+		return {
+			hash: snapshotHash,
+			...snapshot,
```
```suggestion
			...snapshot,
			hash: snapshotHash,
```

azrikahar replied:
I believe the intent (at the time) was to simply follow the "hash at the front" logic for the collections/fields/relations:

https://github.com/directus/directus/blob/776c84c2634b03010ed6e4663876b98f583d92d9/api/src/services/schema.ts#L295-L300

but in hindsight I do think there isn't an actual purpose in keeping it in front at this point!

rijkvanzanten commented on api/src/types/snapshot.ts:
```diff
@@ -13,20 +15,30 @@ export type Snapshot = {
 export type SnapshotField = Field & { meta: Omit<FieldMeta, 'id'> };
 export type SnapshotRelation = Relation & { meta: Omit<RelationMeta, 'id'> };
 
+export type SnapshotWithHash = Omit<Snapshot, 'collections' | 'fields' | 'relations'> & {
```
Do we need the omit here if it's overridden by the second object? I thought the `&` worked as an "assign" instead of a "merge"

azrikahar replied:
This is definitely not my strong suit, but it seems like removing the `Omit` makes typescript unhappy to build it. Any idea why? 🤔 

I believe I might've actually resorted to `Omit` specifically to resolve this at the time, but unfortunately now I honestly don't remember _why_, really sorry for that 😅

azrikahar replied:
Oh yeah that makes it much clearer now. I think that was indeed the original reasoning 👍

rijkvanzanten commented on api/src/types/snapshot.ts:
```diff
@@ -13,20 +15,30 @@ export type Snapshot = {
 export type SnapshotDiff = {
 	collections: {
 		collection: string;
+		hash?: string;
```
Should we make `hash` required everywhere? Looks like it's used quite often 🤔

paescuj replied:
Got superfluous with https://github.com/directus/directus/pull/14713/commits/e70fe8e9547588f41e8d98db2d5d03d4191a1033.

rijkvanzanten commented on api/src/utils/apply-diff.ts:
```diff
@@ -0,0 +1,305 @@
+	diff: Diff<Collection | undefined>[];
+};
+
+export async function applyDiff(
```
This could do with a unit test if at all possible 👀

azrikahar replied:
I believe it is still technically tested by `apply-snapshot.test.ts` since we're extracting `applySnapshot`'s inner logic into `applyDiff`, albeit indirectly now. Should we move it into `apply-diff.test.ts` instead?  🤔 though I'm not sure if `applySnapshot` still requires testing after we do so (I assume that's perfectly fine?).

rijkvanzanten commented on api/src/utils/apply-diff.ts:
```diff
@@ -0,0 +1,305 @@
+				({ diff }) => (diff[0] as DiffDeleted<Collection>).lhs?.meta?.group === currentLevelCollection
+			) as CollectionDelta[];
+
+		const createCollections = async (collections: CollectionDelta[]) => {
+			for (const { collection, diff } of collections) {
+				if (diff?.[0].kind === 'N' && diff[0].rhs) {
+					// We'll nest the to-be-created fields in the same collection creation, to prevent
+					// creating a collection without a primary key
+					const fields = snapshotDiff.fields
+						.filter((fieldDiff) => fieldDiff.collection === collection)
```
I knew this stuff looked familiar 😄 https://github.com/directus/directus/commit/ce104b6a9ccf7b9b1cccf820dbc3f5bbeafe1e94

rijkvanzanten commented on api/src/utils/get-snapshot.ts:
```diff
@@ -61,3 +65,7 @@ function sortDeep(raw: any): any {
 }
+
+export function getVersionedHash(item: Record<string, any>): string {
+	return getSimpleHash(`${JSON.stringify(item)}_${version}`);
```
Alternatively you could use `object-hash` (already a dependency of the API)

rijkvanzanten commented on api/src/utils/get-string-from-stream.ts:
```diff
@@ -0,0 +1,18 @@
+import { Readable } from 'node:stream';
+import { InvalidPayloadException } from '../exceptions';
+
+export function getStringFromStream(stream: Readable) {
```
Needs unit test

rijkvanzanten replied:
Also lets put this in `@directus/utils` (the new utils package)

rijkvanzanten commented on api/src/utils/sanitize-schema.ts:
```diff
@@ -0,0 +1,76 @@
+ */
+
+export function sanitizeCollection(collection: Collection | undefined) {
+	const omittedPaths = [
```
Instead of omitting paths, should we rely on an allow-list of keys that we know we use / rely on and use `pick`?

azrikahar replied:
I think omitting path was first chosen because for Field, the omitting is "simpler" than allowing 😄 but allow-list does seem safer in case there's ever new keys added in knex-schema-inspector side I suppose. Will switch to allow-list 👍

azrikahar committed:
change `were` to `was`

azrikahar committed:
change `where` to `were`

azrikahar commented on api/src/controllers/server.ts:
```diff
@@ -81,4 +89,104 @@ router.get(
+	busboy.on('error', (error: Error) => next(error));
+
+	busboy.on('close', () => {
+		if (fileCount === 0) return next(new InvalidPayloadException(`No file were included in the body`));
```
IIRC this `were` was primarily meant to fix the `where` typo over in files controller:

https://github.com/directus/directus/blob/dc135f776d3ba453e7764fc5471585bcba0ddc18/api/src/controllers/files.ts#L113

but totally did not notice it's outdated after changing `files` to `file` over here 😄

(also updated the typo in files controller after resolving this)

azrikahar committed:
add some empty lines for readability

azrikahar committed:
tweak exception message

azrikahar committed:
fix test

azrikahar committed:
use object-hash

azrikahar committed:
use boolean to check whether file is included

azrikahar committed:
simplify request content type check

azrikahar committed:
throw error when multiple files were uploaded

azrikahar committed:
use nullish coalesce instead of short circuit

paescuj commented on tests-blackbox/routes/server/schema.test.ts:
```diff
@@ -0,0 +1,560 @@
+						version: 1,
+						directus: currentDirectusVersion,
+						// TODO: align vendor names
+						vendor: 'postgres',
```
(commenting here only to point out the TODO)

azrikahar replied:
@licitdev I assume the aligning here means between the blackbox test's vendor list:

https://github.com/directus/directus/blob/d2b5881a1211fdce9091495756f3ff2e1f571ba6/tests-blackbox/common/get-dbs-to-test.ts#L1-L11

against the current list in the API:

https://github.com/directus/directus/blob/d2b5881a1211fdce9091495756f3ff2e1f571ba6/api/src/types/database.ts#L3

because we added validation based on the API's list:

https://github.com/directus/directus/blob/d2b5881a1211fdce9091495756f3ff2e1f571ba6/api/src/services/schema.ts#L18-L20

so `postgres10`, `mysql5` and `sqlite3` would indeed be caught by the valdation (I do recall we had the sqlite vs sqlite3 discussion on discord). Any thoughts on how do we want to address this TODO?

Based on the current values, I did thought about maybe we can simply strip the numbers (eg. `postgres10` would become `postgres`) to satisfy the blackbox test, such as:

```diff
					.send({
						version: 1,
						directus: currentDirectusVersion,
-						// TODO: align vendor names
-						vendor: 'postgres',
+						vendor: vendor.replace(/[0-9]/g, ''),
						collections: [],
						fields: [],
						relations: [],
					})
```

but not sure if that is too naive/hacky of a solution 😄

licitdev replied:
@azrikahar Your solution should work well for now, until there's custom vendors / drivers. 💯  
cc @jaads

azrikahar committed:
Update api/src/services/schema.ts

Co-authored-by: Rijk van Zanten <rijkvanzanten@me.com>

azrikahar commented on api/src/controllers/server.ts:
```diff
@@ -81,4 +89,104 @@ router.get(
 );
 
+router.get(
+	'/schema/snapshot',
```
These are actually simply following the `/server/schema/<operation>` path mentioned in the original issue #11481 😄 but in hindsight I should've questioned it before committing. Definitely agree that `/schema` endpoints makes perfect sense!

paescuj replied:
> These are actually simply following the `/server/schema/<operation>` path mentioned in the original issue #11481 😄 but in hindsight I should've questioned it before committing. Definitely agree that `/schema` endpoints makes perfect sense!

Ah right 👍 Yup, let's change that 😃🚀

paescuj replied:
Done! 👍

paescuj committed:
Remove unnecessary `Omit` on `SnapshotWithHash` type

paescuj commented on api/src/utils/get-snapshot.ts:
```diff
@@ -67,5 +67,5 @@ function sortDeep(raw: any): any {
 
 export function getVersionedHash(item: Record<string, any>): string {
-	return getSimpleHash(`${JSON.stringify(item)}_${version}`);
+	return `${hash(item)}_${version}`;
```
Could we somehow include the version in the hash?

azrikahar replied:
It was definitely an oversight! It should probably be `hash({ ...item, version })` instead 🤔

paescuj replied:
> update: went with hash({ item, version }) for simplicity sake 👍

Exactly how I imagined it 👍 Thanks! ❤️

paescuj committed:
Revert "Remove unnecessary `Omit` on `SnapshotWithHash` type"

This reverts commit d22ac771ec55f07d4b5373204ad803126750f0c9.

azrikahar commented on api/src/database/helpers/schema/dialects/cockroachdb.ts:
```diff
@@ -11,6 +12,17 @@ export class SchemaHelperCockroachDb extends SchemaHelper {
 		await this.changeToTypeByCopy(table, column, type, options);
 	}
 
+	processField(field: Field): void {
```
@licitdev do you recall the reasoning for https://github.com/directus/directus/commit/816c998291ecb1fceb7b7f867f94290896a3435f? I assume it's because knex-schema-inspector doesn't "recognize" this for CockroachDB and set the `has_auto_increment` flag, but also wondering if you have a better comment so wanted to tap you in here 😁

licitdev replied:
Yeah it has been awhile, I recall it being along the lines of not correctly detecting `auto_increment` as the default value from the a string along the lines of `nextval(some_sequence_here)`.

licitdev replied:
Pushed https://github.com/knex/knex-schema-inspector/pull/135. Added a task to update the `knex-schema-inspector` dependency.
Resolving this conversation for now.

paescuj replied:
Awesome, thanks @licitdev! 🚀

paescuj replied:
@licitdev Just to be sure: This means https://github.com/knex/knex-schema-inspector/pull/135 needs to be merged & released before merging this pull request?

licitdev replied:
Yes, that is required.

paescuj replied:
Okay, thanks for your confirmation! Sorry, just saw that you've even added this to the PR's description 🙈

azrikahar commented on api/src/services/schema.ts:
```diff
@@ -0,0 +1,290 @@
+import { getSnapshotDiff } from '../utils/get-snapshot-diff';
+
+const snapshotJoiSchema = Joi.object({
+	version: Joi.number().required(),
```
Sounds good to me! 👍

azrikahar commented on api/src/types/snapshot.ts:
```diff
@@ -1,10 +1,12 @@
 export type Snapshot = {
 	version: number;
 	directus: string;
+	vendor?: DatabaseClient;
```
Yeah, particularly because we also allow omitting the `vendor` flag via `force` query param, so old snapshots can still be used if it's their only option.

paescuj replied:
Great, thanks! 😃 Makes sense 👍

azrikahar committed:
check empty snapshot earlier

azrikahar committed:
use allow-list logic via pick instead of omit

azrikahar committed:
Update api/src/services/schema.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

paescuj committed:
Move to own /schema endpoint controller

paescuj committed:
Fix refs to schema endpoint

azrikahar committed:
move streamToString to utils package

azrikahar committed:
move get-versioned-hash and add test

azrikahar committed:
extract kind into an enum

licitdev committed:
Fix mysql5 timestamp value

licitdev committed:
Fix test collection existing on rerun

azrikahar committed:
resolve TODO comment in blackbox test

paescuj committed:
Drop deep level hashes in diff

These hashes are used only for more accurate error reporting
but are otherwise superfluous, since changes are already
detected by the top level hash.
Therefore we remove them in favor of a simpler diff format and
implementation.

paescuj commented on .gitignore:
```diff
@@ -19,8 +19,8 @@ dist
 TODO
-schema.yaml
-schema.json
+/schema.yaml
+/schema.json
```
Limited to `/` in order to be able to create the file `packages/specs/src/components/schema.yaml`.

@rijkvanzanten Do you remember why this has been added? (https://github.com/directus/directus/pull/8349/commits/fc1c87ac02d2622931defbf4b51cc5ce74042f44) Are those expected in another directory?

rijkvanzanten replied:
It was just so I wouldn't accidentally commit my local schema snapshot debugging attempts 😁

paescuj replied:
I see, thanks 😁 Should we leave it like this or maybe expand it to `/api/schema.yaml`?

rijkvanzanten replied:
The `/` prefix should be fine :

paescuj committed:
Revert schema "fix" for createItem, add comment

paescuj committed:
Strict diff schema validation

paescuj commented on api/src/services/schema.ts:
```diff
@@ -0,0 +1,290 @@
+		),
+		relations: Joi.array().items(
+			Joi.object({
+				hash: Joi.string().when('diff.0.kind', { is: 'N', then: Joi.optional(), otherwise: Joi.any() }),
```
Addressed in aec94e0 ✅

licitdev committed:
Revert CrDB auto-increment detection patch in 816c998

licitdev committed:
Clear systemCache to expose newly added fields

licitdev committed:
Merge branch 'main' into issue/11481

licitdev committed:
Use DiffKind constants

paescuj committed:
Extract diff & snapshot validation into own tested utils

azrikahar commented on api/src/utils/validate-snapshot.ts:
```diff
@@ -0,0 +1,80 @@
+	const { error } = snapshotJoiSchema.validate(snapshot);
+	if (error) throw new InvalidPayloadException(error.message);
+
+	if (!force) {
```
Thoughts on replacing this with `if (force) return;` instead?

azrikahar commented on api/src/utils/validate-diff.ts:
```diff
@@ -0,0 +1,150 @@
+		return false;
+	}
+
+	if (applyDiff.hash !== currentSnapshotWithHash.hash) {
```
Similar to validateSnapshot, how do we feel with doing `if (applyDiff.hash === currentSnapshotWithHash.hash) return false;` instead? (or `if (applyDiff.hash === currentSnapshotWithHash.hash) return;` depending on our discussion in https://github.com/directus/directus/pull/14713#discussion_r1090691487)

paescuj committed:
Apply suggestions from @azrikahar

paescuj committed:
Merge branch 'main' into issue/11481

paescuj committed:
Update knex-schema-inspector to 3.0.1

Includes the fix for CrDB auto-increment detection (knex/knex-schema-inspector#135)

licitdev committed:
Update knex-schema-inspector in packages

licitdev committed:
Update lock file

azrikahar committed:
add test for schema service

azrikahar committed:
add test for export service

azrikahar committed:
add relevant tests to modified util functions

azrikahar committed:
fix csv test to account for os end of line

azrikahar committed:
fix files controller test

paescuj commented on api/src/services/schema.test.ts:
```diff
@@ -0,0 +1,331 @@
+describe('Services / Schema', () => {
+	describe('snapshot', () => {
+		it('should throw ForbiddenException for non-admin user', async () => {
+			const testSnapshot = {
```
Can we dedup all those test data (in this case with line 59)? Same with `snapshotDiffWithHash`, ...

azrikahar replied:
I did thought about it, but there was some discussion on making tests more "contained" within themselves (even if it results in dupes) 🤔 That being said, it's probably less important here so I'll proceed to dedupe them 👍

paescuj replied:
Okay, I see! Whatever feels right for you 👍😃

azrikahar replied:
Done! 🙌

paescuj replied:
Nice 🙌  Thanks!

azrikahar committed:
dedupe test data for schema service

paescuj committed:
Align schema specs with docs

rijkvanzanten committed:
Merge branch 'main' into issue/11481

rijkvanzanten commented on api/src/controllers/schema.ts:
```diff
@@ -0,0 +1,120 @@
+
+const schemaMultipartHandler: RequestHandler = (req, res, next) => {
+	if (req.is('application/json')) {
+		if (Object.keys(req.body).length === 0) throw new InvalidPayloadException(`No data were included in the body`);
```
```suggestion
		if (Object.keys(req.body).length === 0) throw new InvalidPayloadException(`No data was included in the body`);
```

rijkvanzanten commented on api/src/controllers/schema.ts:
```diff
@@ -0,0 +1,120 @@
+				'content-type': 'application/octet-stream',
+		  };
+
+	const busboy = Busboy({ headers });
```
Lets add `limits: { fileSize }` here with a env var for max schema size to process. Related to ENG-491

rijkvanzanten replied:
Alternatively we could hardcode it to a safe upper limit (like 5MB), but this feels like something that should be configurable through env

rijkvanzanten replied:
Actually lets combine this with ENG-491 and just make it a single config option for both 👍🏻

rijkvanzanten committed:
Update api/src/controllers/schema.ts

rijkvanzanten commented:
Looks great, feels great, works great, great job everybody 👏🏻

licitdev committed:
Revert testing for all vendors

rijkvanzanten commented:
@licitdev You're the last to approve! 

<img width="672" alt="CleanShot 2023-01-31 at 12 30 48@2x" src="https://user-images.githubusercontent.com/9141017/215837857-043ca467-6e3b-47e9-a4e0-1d5eca24f2f0.png">

RobinTHOREL commented:
hey guys, this feature look like awesome and really exepected. can we find docs somewhere? 👀
